data/transactions/logic/assembler.go

```diff
@@ -248,8 +249,55 @@ type OpStream struct {
 	HasStatefulOps bool
 }
 
-// createLabel inserts a label reference to point to the next
-// instruction, reporting an error for a duplicate.
+// ProgramKnowledge tracks statically known information as we assemble
+type ProgramKnowledge struct {
+	// list of the types known to be on the value stack, based on specs of
+	// opcodes seen while assembling. In normal code, the tip of the stack must
+	// match the next opcode's Arg.Types, and is then replaced with its
+	// Return.Types. If the top entry is StackNone, we're in dead code, and no
+	// checking is done, nor are return types appended.
+	stack StackTypes
+	// bottom is the type given out when known is empty. It is StackNone at
+	// program start, so, for example, a `+` opcode at the start of a program
+	// fails. But when a label or callsub is encountered, `stack` is truncated
+	// and `bottom` becomes StackAny, because we don't track program state
+	// coming in from elsewhere. A `+` after a label succeeds, because the stack
+	// "vitually" contains an infinite list of StackAny.
+	bottom StackType
+}
+
+func (pgm *ProgramKnowledge) pop() StackType {
+	if len(pgm.stack) == 0 {
+		return pgm.bottom
+	}
+	last := len(pgm.stack) - 1
+	t := pgm.stack[last]
+	pgm.stack = pgm.stack[:last]
+	return t
+}
+
+func (pgm *ProgramKnowledge) push(types ...StackType) {
+	pgm.stack = append(pgm.stack, types...)
+}
+
+func (pgm *ProgramKnowledge) deaden() {
+	pgm.stack = append(pgm.stack, StackNone)
+}
+
+// label resets knowledge to reflect that control may enter from elsewhere.
+func (pgm *ProgramKnowledge) label() {
+	if pgm.isDead() {
+		pgm.stack = nil
+		pgm.bottom = StackAny
+	}
+}
+
+func (pgm ProgramKnowledge) isDead() bool {
+	return len(pgm.stack) > 0 && pgm.stack[len(pgm.stack)-1] == StackNone
+}
+
+// createLabel inserts a label to point to the next instruction, reporting an
+// error for a duplicate.
 func (ops *OpStream) createLabel(label string) {
 	if ops.labels == nil {
```

Comment:
@jannotti Minor question predating the PR:  I'm wondering if there's a by-design reason to retain the `ops.labels` initialization here rather than at start of `assemble`?

There seems to be several state initializations (e.g. `ops.OffsetToLine` in `recordSourceLine`) and inline initialization at start of `assembly` (e.g. `ops.typeTracking = true`).

I'm imagining it can be refactored into a single function and called at start of `assemble`.  Before working on the change, looking to sanity check if you see obvious concerns.

Reply:
I think it would be pretty reasonable to do what you say.  Here's how it happened.  There's a Go convention that says you should try to make empty values (here `OpStream{}`), useful. This sort of falls out of Go's lack of true constructors.  You want `OpStream{}` to be usable without much ceremony, if possible.  So recordSourceLine and createLabel were written to perform their own initialization.  But when I introduced `typeTracking` I couldn't bring myself to reverse the meaning (i.e use `skipTypeTracking`) so that false (the "zero" value") would have the default meaning I wanted.  So I had to do some initialization.  Given that, perhaps we should just admit it, introduce NewOpStream(), and do the other initialization there as well.  But then I also want to rename OpStream to Assembly someday.  Is this that day?
I suppose those are unneeded allocations for very simply programs with no labels.  But I have no objection.

Reply:
@jannotti The rename (at least in the current PR) feels bold to me.  But, today _can_ be the day to add `NewOpStream`.  I opened https://github.com/jannotti/go-algorand/pull/7 for your review.

If you find it's more trouble than it's worth, I'll happily close and we can revisit another time.

Reply:
merged